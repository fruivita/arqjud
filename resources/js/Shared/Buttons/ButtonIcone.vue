<!--
    Botão com ícone apenas, isto é, sem texto.

    @link https://vuejs.org/guide/introduction.html
    @link https://tailwindcss.com/docs
    @link https://inertiajs.com/
 -->

<script setup>
import Icone from '@/Shared/Icones/Icone.vue';

defineProps({
    icone: { type: String, required: true },
    especie: {
        type: String,
        default: 'acao',
        validator(value) {
            return ['acao', 'alerta', 'perigo', 'padrao'].includes(value);
        },
    },
    type: {
        type: String,
        default: 'button',
        validator(value) {
            return ['button', 'submit'].includes(value);
        },
    },
});
</script>

<template>
    <button
        :class="{
            'bg-green-500 hover:bg-green-600 hover:ring-green-900 focus:bg-green-600 focus:ring-green-900':
                especie === 'padrao',
            'bg-blue-500 hover:bg-blue-600 hover:ring-blue-900 focus:bg-blue-600 focus:ring-blue-900':
                especie === 'acao',
            'bg-yellow-600 hover:bg-yellow-700 hover:ring-yellow-900 focus:bg-yellow-700 focus:ring-yellow-900':
                especie === 'alerta',
            'bg-red-500 hover:bg-red-600 hover:ring-red-900 focus:bg-red-600 focus:ring-red-900':
                especie === 'perigo',
        }"
        :type="type"
        class="transform-gpu rounded p-2 text-white outline-none transition duration-300 hover:ring-4 focus:ring-4"
        icone="person"
    >
        <Icone :nome="icone" />
    </button>
</template>
