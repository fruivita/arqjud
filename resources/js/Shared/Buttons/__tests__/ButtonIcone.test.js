/**
 * Testes para o componente ButtonIcone.
 *
 * @link https://vuejs.org/guide/scaling-up/testing.html
 * @link https://test-utils.vuejs.org/guide/
 * @link https://vitest.dev/
 */

import ButtonIcone from '@/Shared/Buttons/ButtonIcone.vue';
import { createTestingPinia } from '@pinia/testing';
import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, test } from 'vitest';

let mountFunction;
let pinia;

beforeEach(() => {
    pinia = (options = {}) =>
        createTestingPinia({
            initialState: {
                StatusRequisicaoStore: { ...options },
            },
        });

    mountFunction = (options = {}) => {
        return mount(ButtonIcone, { ...options });
    };
});

describe('ButtonIcone', () => {
    // Caminho feliz
    const types = ['button', 'submit'];
    const especies = ['acao', 'alerta', 'perigo', 'padrao'];

    test('aceita apenas os types definidos', () => {
        const validator = ButtonIcone.props.type.validator;

        types.forEach((type) => expect(validator(type)).toBe(true));
        expect(validator('foo')).toBe(false);
    });

    test('aceita apenas as espécies definidas', () => {
        const validator = ButtonIcone.props.especie.validator;

        especies.forEach((especie) => expect(validator(especie)).toBe(true));
        expect(validator('foo')).toBe(false);
    });

    test('propriedades do componente estão definidas', () => {
        expect(ButtonIcone.props).toMatchObject({
            icone: { type: String, required: true },
            especie: { type: String, default: 'acao' },
            type: { type: String, default: 'button' },
        });
    });

    test.each(types)(
        'renderiza o componente com os types específicos respeitando o snapshot',
        (type) => {
            expect(
                mountFunction({
                    props: { icone: 'key', type: type },
                    global: {
                        plugins: [pinia({ processando: false })],
                    },
                }).html()
            ).toMatchSnapshot();
        }
    );

    test.each(especies)(
        'renderiza o componente das espécies específicas respeitando o snapshot',
        (especie) => {
            expect(
                mountFunction({
                    props: { icone: 'key', especie: especie },
                    global: {
                        plugins: [pinia({ processando: false })],
                    },
                }).html()
            ).toMatchSnapshot();
        }
    );

    test('renderiza o componente de maneira diversa enquanto a página se carrega respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: { icone: 'key' },
                global: {
                    plugins: [pinia({ processando: true })],
                },
            }).html()
        ).toMatchSnapshot();
    });
});
