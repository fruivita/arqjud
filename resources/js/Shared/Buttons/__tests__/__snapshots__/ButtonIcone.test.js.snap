// Vitest Snapshot v1

exports[`ButtonIcone > renderiza o componente com os types específicos respeitando o snapshot 1`] = `
"<button class=\\"bg-blue-500 hover:bg-blue-600 hover:ring-blue-900 focus:bg-blue-600 focus:ring-blue-900 transform-gpu rounded p-2 text-white outline-none transition duration-300 hover:ring-4 focus:ring-4\\" type=\\"button\\" icone=\\"person\\"><svg class=\\"w-6 h-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
    <path d=\\"M0 8a4 4 0 0 1 7.465-2H14a.5.5 0 0 1 .354.146l1.5 1.5a.5.5 0 0 1 0 .708l-1.5 1.5a.5.5 0 0 1-.708 0L13 9.207l-.646.647a.5.5 0 0 1-.708 0L11 9.207l-.646.647a.5.5 0 0 1-.708 0L9 9.207l-.646.647A.5.5 0 0 1 8 10h-.535A4 4 0 0 1 0 8zm4-3a3 3 0 1 0 2.712 4.285A.5.5 0 0 1 7.163 9h.63l.853-.854a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.793-.793-1-1h-6.63a.5.5 0 0 1-.451-.285A3 3 0 0 0 4 5z\\"></path>
    <path d=\\"M4 8a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
  </svg></button>"
`;

exports[`ButtonIcone > renderiza o componente com os types específicos respeitando o snapshot 2`] = `
"<button class=\\"bg-blue-500 hover:bg-blue-600 hover:ring-blue-900 focus:bg-blue-600 focus:ring-blue-900 transform-gpu rounded p-2 text-white outline-none transition duration-300 hover:ring-4 focus:ring-4\\" type=\\"submit\\" icone=\\"person\\"><svg class=\\"w-6 h-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
    <path d=\\"M0 8a4 4 0 0 1 7.465-2H14a.5.5 0 0 1 .354.146l1.5 1.5a.5.5 0 0 1 0 .708l-1.5 1.5a.5.5 0 0 1-.708 0L13 9.207l-.646.647a.5.5 0 0 1-.708 0L11 9.207l-.646.647a.5.5 0 0 1-.708 0L9 9.207l-.646.647A.5.5 0 0 1 8 10h-.535A4 4 0 0 1 0 8zm4-3a3 3 0 1 0 2.712 4.285A.5.5 0 0 1 7.163 9h.63l.853-.854a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.793-.793-1-1h-6.63a.5.5 0 0 1-.451-.285A3 3 0 0 0 4 5z\\"></path>
    <path d=\\"M4 8a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
  </svg></button>"
`;

exports[`ButtonIcone > renderiza o componente das espécies específicas respeitando o snapshot 1`] = `
"<button class=\\"bg-blue-500 hover:bg-blue-600 hover:ring-blue-900 focus:bg-blue-600 focus:ring-blue-900 transform-gpu rounded p-2 text-white outline-none transition duration-300 hover:ring-4 focus:ring-4\\" type=\\"button\\" icone=\\"person\\"><svg class=\\"w-6 h-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
    <path d=\\"M0 8a4 4 0 0 1 7.465-2H14a.5.5 0 0 1 .354.146l1.5 1.5a.5.5 0 0 1 0 .708l-1.5 1.5a.5.5 0 0 1-.708 0L13 9.207l-.646.647a.5.5 0 0 1-.708 0L11 9.207l-.646.647a.5.5 0 0 1-.708 0L9 9.207l-.646.647A.5.5 0 0 1 8 10h-.535A4 4 0 0 1 0 8zm4-3a3 3 0 1 0 2.712 4.285A.5.5 0 0 1 7.163 9h.63l.853-.854a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.793-.793-1-1h-6.63a.5.5 0 0 1-.451-.285A3 3 0 0 0 4 5z\\"></path>
    <path d=\\"M4 8a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
  </svg></button>"
`;

exports[`ButtonIcone > renderiza o componente das espécies específicas respeitando o snapshot 2`] = `
"<button class=\\"bg-yellow-600 hover:bg-yellow-700 hover:ring-yellow-900 focus:bg-yellow-700 focus:ring-yellow-900 transform-gpu rounded p-2 text-white outline-none transition duration-300 hover:ring-4 focus:ring-4\\" type=\\"button\\" icone=\\"person\\"><svg class=\\"w-6 h-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
    <path d=\\"M0 8a4 4 0 0 1 7.465-2H14a.5.5 0 0 1 .354.146l1.5 1.5a.5.5 0 0 1 0 .708l-1.5 1.5a.5.5 0 0 1-.708 0L13 9.207l-.646.647a.5.5 0 0 1-.708 0L11 9.207l-.646.647a.5.5 0 0 1-.708 0L9 9.207l-.646.647A.5.5 0 0 1 8 10h-.535A4 4 0 0 1 0 8zm4-3a3 3 0 1 0 2.712 4.285A.5.5 0 0 1 7.163 9h.63l.853-.854a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.793-.793-1-1h-6.63a.5.5 0 0 1-.451-.285A3 3 0 0 0 4 5z\\"></path>
    <path d=\\"M4 8a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
  </svg></button>"
`;

exports[`ButtonIcone > renderiza o componente das espécies específicas respeitando o snapshot 3`] = `
"<button class=\\"bg-red-500 hover:bg-red-600 hover:ring-red-900 focus:bg-red-600 focus:ring-red-900 transform-gpu rounded p-2 text-white outline-none transition duration-300 hover:ring-4 focus:ring-4\\" type=\\"button\\" icone=\\"person\\"><svg class=\\"w-6 h-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
    <path d=\\"M0 8a4 4 0 0 1 7.465-2H14a.5.5 0 0 1 .354.146l1.5 1.5a.5.5 0 0 1 0 .708l-1.5 1.5a.5.5 0 0 1-.708 0L13 9.207l-.646.647a.5.5 0 0 1-.708 0L11 9.207l-.646.647a.5.5 0 0 1-.708 0L9 9.207l-.646.647A.5.5 0 0 1 8 10h-.535A4 4 0 0 1 0 8zm4-3a3 3 0 1 0 2.712 4.285A.5.5 0 0 1 7.163 9h.63l.853-.854a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.793-.793-1-1h-6.63a.5.5 0 0 1-.451-.285A3 3 0 0 0 4 5z\\"></path>
    <path d=\\"M4 8a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
  </svg></button>"
`;

exports[`ButtonIcone > renderiza o componente das espécies específicas respeitando o snapshot 4`] = `
"<button class=\\"bg-green-500 hover:bg-green-600 hover:ring-green-900 focus:bg-green-600 focus:ring-green-900 transform-gpu rounded p-2 text-white outline-none transition duration-300 hover:ring-4 focus:ring-4\\" type=\\"button\\" icone=\\"person\\"><svg class=\\"w-6 h-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
    <path d=\\"M0 8a4 4 0 0 1 7.465-2H14a.5.5 0 0 1 .354.146l1.5 1.5a.5.5 0 0 1 0 .708l-1.5 1.5a.5.5 0 0 1-.708 0L13 9.207l-.646.647a.5.5 0 0 1-.708 0L11 9.207l-.646.647a.5.5 0 0 1-.708 0L9 9.207l-.646.647A.5.5 0 0 1 8 10h-.535A4 4 0 0 1 0 8zm4-3a3 3 0 1 0 2.712 4.285A.5.5 0 0 1 7.163 9h.63l.853-.854a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.793-.793-1-1h-6.63a.5.5 0 0 1-.451-.285A3 3 0 0 0 4 5z\\"></path>
    <path d=\\"M4 8a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
  </svg></button>"
`;
